### ISSUE TITLE
RUBI integrator infrastructure

### ISSUE THREAD

**Issue by certik** [symengine/symengine#1719]
@naveensaigit and I brainstormed steps that could be done to get the infrastructure for rule rewriting and decision tree generation working in SymEngine. Our goal is to start with a few rules, but get them working all the way, and then keep adding more rules.

I will be updating the issue description with more details.

**Comment by certik** [comment]
The branch `w` and this file: https://github.com/certik/rubi_generated/blob/79613660c77db95736b1d4eb6fbfb29081230364/run_example.py contains a test for a few simple rules that seem to work.

**Comment by certik** [comment]
@Upabjojr, do you know if the decision tree generated could be modified to output something like:
```python
def generated_rule(b, n, x):
    if (ZeroQ(n+S(1))):      
        return replacement2(b, n, x)
    else:
        return replacement1(b, n, x)
```
instead of the current output, such as in https://github.com/certik/rubi_generated/blob/79613660c77db95736b1d4eb6fbfb29081230364/gen/generated_part000001.py, which seems really complex and probably slower?

**Comment by Upabjojr** [comment]
> @Upabjojr, do you know if the decision tree generated could be modified to output something like:

Do you mean the decision tree generated by MatchPy? The code follows the exact way in which MatchPy operates internally, so it would be quite tough to customize the generated code.

**Comment by certik** [comment]
Yes, the tree generated by the file https://github.com/Upabjojr/rubi_generated/blob/cd35e9e51722b04fb159ada3d5811d62a423e429/decision_tree_rubi_generator.py. I thought you wrote it.

**Comment by Upabjojr** [comment]
It's subclassing _MatchPy_'s `CodeGenerator` object in order to do some customizations of the generated code. 

> ```python
> def generated_rule(b, n, x):
>     if (ZeroQ(n+S(1))):      
>         return replacement2(b, n, x)
>     else:
>         return replacement1(b, n, x)
> ```

Where is this code coming from?

**Comment by certik** [comment]
That's the code that I would like to get generated. If it can be generated, together with the pattern, then it will greatly simplify the decision tree and also probably make it faster.

**Comment by Upabjojr** [comment]
Well, I'm not sure it can be generated like that because it is missing all the logic to match the elements of the expression tree.

**Comment by Upabjojr** [comment]
By the way, I feel that the decision tree is being generated more complicated than what it ought be. Incidentally, I believe it's an issue of replication of objects in the rule definition.

The rules may not be optimal when they are added to MatchPy.

**Comment by certik** [comment]
@Upabjojr would you have time to do a video call with @naveensaigit and I? We have been working on this lately. We think it should be possible to generate, but we have to write the generator, either using MatchPy or even from scratch.

**Comment by Upabjojr** [comment]
> @Upabjojr would you have time to do a video call with @naveensaigit and I? We have been working on this lately. We think it should be possible to generate, but we have to write the generator, either using MatchPy or even from scratch.

Yes, sure!

**Comment by Upabjojr** [comment]
Please remember that unless there is support for the coroutines introduced in C++20, any attempt at using RUBI in SymEngine will either timeout or cause a memory error.

**Comment by rikardn** [comment]
I implemented the first two patterns of Rubi-5 in https://github.com/symengine/symengine/pull/1801.

**Comment by Upabjojr** [comment]
Maybe we need a way to automatically translate the code into C++.

**Comment by Upabjojr** [comment]
> (only 42 rules as compared to 3000 for Rubi-4)

Have you tried to use the submodule `matchPyCpp` on those 42 rules? It's a Python submodule in SymEngine that is able to transform Mathematica pattern matching into a structure of nested if-/for-loops in C++.

**Comment by Upabjojr** [comment]
Also, as a reminder, this work could be useful to SymPy as well...

**Comment by rikardn** [comment]
Thanks for your input @Upabjojr! It's a good idea to try `matchPyCpp`. Since it was "only" 42 rules I was think they could be translated manually, but I might be naïve. The thousands of if/then/else functions must still be machine translated. This translation could perhaps also generate code for SymPy.

(We also have to keep in mind that Rubi-5 is far from ready.)
